docs/ZipperTutorial.ml

```diff
@@ -0,0 +1,78 @@
+(* An introduction to top-down zippers for structured editing *)
+
+(* Given an algebraic data type, we can mechanically generate another type,
+   typically called the zipper or the type of 'one-holed contexts' relative
+   to the original type. Zippers come in top-down and bottom-up varieties.
+   Here we focus on top-down zippers, the variant used in Hazel.
+
+   Conceptually a value of this new type carries the same information
+   as a corresponding value of the original type, except that it also
+   distinguishes a particular subterm, which we call the 'cursor term'.
+
+   Given a zippered type, we can also mechanically define a corresponding
+   type of movement actions, each of which coresponds to a way of locally
+   changing which subterm is being indicated. *)
+
+(* A simplified AST type, representing unannotated binary trees *)
+type ast =
+  | Atom
+  | Bin of ast * ast
+
+(* A zippered representation of the above AST. In general, the zippered
+   version of a recursive ADT will have n variants for every n-ary
+   constructor (n-ary in the sense of containing n references to the
+   recursive type), plus an aditional variant representing the cursor
+   itself. So in our example, we have 0 variants for the 0-ary Atom,
+   2 variants for the binary Bin (representing the cursor being within
+   the left or right subtrees respectively), and the cursor case,
+   giving 3 variants in total *)
+type zast =
+ | Cursor of ast
+ | BinL of zast * ast
+ | BinR of ast * zast
+
+ (* Simplified local movement actions on the zippered AST, representing either moving
+    the cursor up towards the root, or down the left or right branch of a Bin *)
+ type action = 
+ | SelectParent
+ | SelectChildBinLeft
+ | SelectChildBinRight
+
+ (* To get a handle on the nature of the zippered representation, implement the
+    following four functions. In doing so, you will have created almost all the
+    elements of a simple structured editor using a model-view-update architecture! *)
+
+(* 1. Convert from a non-zippered ast to zippered ast by selecting the root. Note
+      that the correspondance in the direction ast -> zast is one-to-many; each
+      tree has many possible cursor placements. Here we just want to do the
+      simplest possible thing; in fact, the unique thing which can be done
+      without inspecting the ast structure *)
+ let select_root : ast -> zast = failwith "TODO"
+;; assert (select_root (Bin (Atom, Atom)) = Cursor (Bin (Atom, Atom)))
+
+(* 2. Convert from zippered ast to non-zippered ast by forgetting about the cursor.
+      Note that unlike the other direction, each zippered value corresponds to
+      exactly one non-zippered value *)
+let erase : zast -> ast = failwith "TODO"
+;; assert (erase (BinL (Cursor Atom, Atom)) = Bin (Atom, Atom))
+
+(* 3. Simplified view function: Print a zast, using a caret (>) to indicate the cursor term *)
```

Comment:
these tests could be the examples, and maybe have this be the first exercise in that case

Reply:
i feel the flow makes more sense starting with selection rather than erasure; added examples earlier tho